src/domain/entities/Questionnaire/Questionnaire.ts

```diff
@@ -0,0 +1,178 @@
+import { Id, Ref } from "../Ref";
+import _ from "../generic/Collection";
+import { Code, Question } from "./QuestionnaireQuestion";
+import { QuestionnaireRule, getApplicableRules } from "./QuestionnaireRules";
+import { QuestionnaireSection, QuestionnaireSectionM } from "./QuestionnaireSection";
+
+export interface QuestionnaireBase {
+    id: Id;
+    name: string;
+    description: string;
+    orgUnit: Ref;
+    year: string;
+    isCompleted: boolean;
+    isMandatory: boolean;
+}
+
+interface QuestionnaireData extends QuestionnaireBase {
+    stages: QuestionnaireStage[];
+    entity?: QuestionnaireEntity; //Equivalant to tracked entity instance of tracker program
+    subLevelDetails?: {
+        enrollmentId: Id;
+    };
+    rules: QuestionnaireRule[];
+}
+
+export interface QuestionnaireEntity {
+    title: string;
+    code: string;
+    questions: Question[];
+    isVisible: boolean;
+    stageId: string;
+}
+
+export interface QuestionnaireStage {
+    title: string;
+    code: Code;
+    sections: QuestionnaireSection[];
+    isVisible: boolean;
+    showNextStage?: boolean;
+    instanceId?: Id; //Corresponds to DHIS eventId
+}
+
+export class Questionnaire {
+    private readonly data: QuestionnaireData;
+
+    private constructor(data: QuestionnaireData) {
+        this.data = data;
+    }
+    public get stages(): QuestionnaireStage[] {
+        return this.data.stages;
+    }
+
+    public get rules(): QuestionnaireRule[] {
+        return this.data.rules;
+    }
+    public get entity(): QuestionnaireEntity | undefined {
+        return this.data.entity;
+    }
+    public get orgUnit(): Ref {
+        return this.data.orgUnit;
+    }
+
+    public get subLevelDetails(): { enrollmentId: Id } | undefined {
+        return this.data.subLevelDetails;
+    }
+
+    public static create(data: QuestionnaireData): Questionnaire {
+        //TO DO : Add validations if any
+        return new Questionnaire({
+            id: data.id,
+            name: data.name,
+            description: data.description,
+            orgUnit: data.orgUnit,
+            year: data.year,
+            isCompleted: data.isCompleted,
+            isMandatory: data.isMandatory,
+            stages: data.stages,
+            entity: data.entity,
+            subLevelDetails: data.subLevelDetails,
+            rules: data.rules,
+        });
+    }
+
+    static updateQuestionnaireEntity(
+        questionnaire: Questionnaire,
```

Comment:
Now that we have a class, all static methods which were taking a Questionnare can be converted to instance methods.